blockchain/utxocache.go

```diff
@@ -0,0 +1,816 @@
+	if statusCode == ucsConsistent && *statusHash == tip.hash {
+		log.Debugf("UTXO state consistent (%d:%v)", tip.height, tip.hash)
+		return nil
+	}
+
+	log.Info("Reconstructing UTXO state after unclean shutdown. This may take " +
+		"a long time...")
+
+	// Even though this should always be true, make sure the fetched hash is in
+	// the best chain.
+	var statusNode *blockNode
+	var statusNodeNext *blockNode // the first one higher than the statusNode
+	for node := tip; node.height > 0; node = node.parent {
+		if node.hash == *statusHash {
+			statusNode = node
+			break
+		}
+		statusNodeNext = node
+	}
+	if statusNode == nil {
+		return AssertError(fmt.Sprintf("last utxo consistency status contains "+
+			"hash that is not in best chain: %v", statusHash))
+	}
+
+	// If data was in the middle of a flush, we have to roll back all blocks from
+	// the last best block all the way back to the last consistent block.
+	if statusCode == ucsFlushOngoing {
+		log.Debugf("btcd was shut down during a UTXO cache flush, "+
+			"rolling back %d blocks...", tip.height-statusNode.height)
+
+		// Roll back blocks in batches.
+		rollbackBatch := func(dbTx database.Tx, node *blockNode) (*blockNode, error) {
+			nbBatchBlocks := 0
+			for ; node.height > statusNode.height; node = node.parent {
+				block, err := dbFetchBlockByNode(dbTx, node)
+				if err != nil {
+					return nil, err
+				}
+
+				stxos, err := dbFetchSpendJournalEntry(dbTx, block)
+				if err != nil {
+					return nil, err
+				}
+
+				if err := s.rollBackBlock(block, stxos); err != nil {
+					return nil, err
+				}
+				nbBatchBlocks++
+
+				if nbBatchBlocks >= utxoBatchSizeBlocks {
+					break
+				}
+			}
+			return node, nil
+		}
+		for node := tip; node.height > statusNode.height; {
+			log.Tracef("Rolling back %d more blocks...",
+				node.height-statusNode.height)
+			err := s.db.Update(func(dbTx database.Tx) error {
+				var err error
+				node, err = rollbackBatch(dbTx, node)
+				return err
+			})
+			if err != nil {
+				return err
+			}
+
+			if interruptRequested(interrupt) {
+				log.Warn("UTXO state reconstruction interrupted")
+				return errInterruptRequested
+			}
+		}
+
+		// Now we can update the status already to avoid redoing this work when
+		// interrupted.
+		err := s.db.Update(func(dbTx database.Tx) error {
+			return dbPutUtxoStateConsistency(dbTx, ucsConsistent, statusHash)
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	log.Debugf("Replaying %d blocks to rebuild UTXO state...",
+		tip.height-statusNodeNext.height)
+
+	// Then we replay the blocks from the last consistent state up to the best
+	// state.  Iterate forward from the consistent node to the tip of the best
+	// chain.  After every batch, we can also update the consistency state to
+	// avoid redoing the work when interrupted.
+	rollforwardBatch := func(dbTx database.Tx, node *blockNode) (*blockNode, error) {
+		nbBatchBlocks := 0
+		for ; node.height <= tip.height; node = node.parent {
```

Comment:
you are rolling back nodes here rather than rolling them forward. 

Should be something more like:
```
for e := attachNodes.Front(); e != nil; e = e.Next() {
			node = e.Value.(*blockNode)
			attachNodes.Remove(e)
```